hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieFileGroupReaderSchemaHandler.java

```diff
@@ -162,6 +165,31 @@ private Schema generateRequiredSchema() {
     return appendFieldsToSchemaDedupNested(requestedSchema, addedFields);
   }
 
+  private static String[] getMandatoryFieldsForMerging(HoodieTableConfig cfg, Option<HoodieRecordMerger> recordMerger) {
+    if (cfg.getRecordMergeMode() == RecordMergeMode.CUSTOM) {
+      return recordMerger.get().getMandatoryFieldsForMerging(cfg);
```

Comment:
Should the default implementation of `getMandatoryFieldsForMerging` return all columns since custom merging may need all columns regardless of the request schema (to be followed up in a separate fix)?  Only the `DefaultSparkRecordMerger` and `OverwriteWithLatestSparkRecordMerger`(and counterpart classes for other engines) should overwrite the implementation to return record key and optional preCombine field.

Reply:
Let's create a follow-up Jira ticket on this since it requires more work to make the usability better.  Right now the user must provide `getMandatoryFieldsForMerging` to make sure the functionality is correct, which is OK as long as we update the docs (let's add a Jira ticket too).

Reply:
HUDI-8420 and HUDI-8421 as follow-ups.